### ISSUE TITLE
快速构建一个cnn神经网络

### ISSUE THREAD

**Issue by Hmy-bilibili** [Hmy-bilibili/-cnn-#1]
0.前言
每当想要学习神经网络的时候，每当想要自己搭建一个神经网络的时候，每当看到别人的神经网络错综复杂就劝退的时候，不如尝试一下搭建一个最简单的CNN试试。或许会有那种搭积木的感觉。

1.mnist手写数字数据集
mnist手写数字数据集是一系列的28x28x1的灰度图片，它在神经网络里面的地位相当于编程里面的hello world。

2.网络结构
1)输出层
我们经常把神经网络比作黑盒模型，那么黑盒就要有一个输入与输出，我们知道MNIST手写数字是一张张28*28像素的图片，那么输出就是对应的0-9的10个标签数字。
如果直接输出0-9，那么就会出现数字之间的相关性，比如1后面是2，在生成的神经网络的权重就会带有数字的顺序特征，所以一般的输出采用的是one-hot模式，就是用一个1维的矩阵来表示输出。
例如：
3：[0 0 0 1 0 0 0 0 0 0]
5：[0 0 0 0 0 1 0 0 0 0]
各个输出矩阵之间相互正交，乘积为0。
最后会用一个全连接层来将卷积运算之后的矩阵映射到one-hot上，并使用softmax将输出转换为概率。

2)中间层(隐藏层)
中间部分就是卷积层以及池化层，这两个层不断的叠加以及传递数据，卷积层就是用带加权值的卷积核提取图片特征，而池化层的作用则是减小处理后矩阵的输出大小和降低过拟合。
大致的结构就是
卷积->池化->卷积->池化->卷积->全连接

3.具体代码
理论结束，接下来用tensorflow实现
1)导入并处理数据
2)构建网络模型，并运行网络

3)验证模型

4.网络输出
1)网络结构

2)网络运行结果

5.总结
可以看到，结果还算一般，准确率有98.98%，多做几个epochs效果应该会更好。使用Keras来搭建模型确实会轻松很多，而且代码可以跨多版本的tensorflow。
总的流程就是 处理数据->搭建模型->选择合适的优化器，激活函数 ->输出结果
祝大家学习愉快。
``
from tensorflow.keras.datasets import mnist
from tensorflow.keras.layers import Dense, Conv2D, MaxPooling2D, Flatten
from tensorflow.keras.models import Sequential
from tensorflow.keras.utils import to_categorical

(x_train, y_train), (x_test, y_test) = mnist.load_data()   #加载mnist数据

x_train = x_train.reshape((-1, 28, 28, 1))   
#重新变换数组的shape属性，其中-1表示自动计算大小，这里的-1是指x_train的数量(60000)

x_train = x_train.astype('float32')/255        
#将数据转换为32位浮点型，并归一化处理

y_train = to_categorical(y_train)              
#to_categorical就是将类别向量转换为二进制（只有0和1）的矩阵类型表示。
#其表现为将原有的类别向量转换为独热编码的形式。

x_test = x_test.reshape((-1, 28, 28, 1))
x_test = x_test.astype('float32')/255
y_test = to_categorical(y_test)
#以上同理

#再将训练集中的前50000个划为训练集，后10000个作为验证集。
x_val = x_train[:10000]
y_val = y_train[:10000]
partial_x_train = x_train[10000:]
partial_y_train = y_train[10000:]

model = Sequential()
#声明网络为顺序模型，这种模型是单输入单输出的模式，层之间只有相邻关系，不存在跨层连接
#还有一种模型是Model，多输入多输出，层与层之间任意连接。这种模型编译速度慢。
model.add(Conv2D(32, (3, 3), padding='same', activation='relu', input_shape=(28, 28, 1)))
#添加一个卷积层，padding声明为same，表示矩阵经过卷积之后，形状不变，激活函数为relu
model.add(MaxPooling2D((2, 2)))
#添加一个池化层
model.add(Conv2D(64, (3, 3),padding='same', activation='relu'))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(64, (3, 3), padding='same', activation='relu'))
model.add(Flatten())
#添加一个过度层，将卷积层的结果压扁层一个一维矩阵
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))
#添加全连接层，并通过softmax输出概率
model.summary()
#这个函数是将以上的神经网络输出

model.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])
#optimizer优化器选择adam
#loss损失函数选择categorical_crossentropy

model.fit(partial_x_train, partial_y_train, validation_data=(x_val, y_val), epochs=20, batch_size=64)
#导入训练集以及验证集，开始训练模型

test_loss, test_acc = model.evaluate(x_test, y_test)
print(str(test_acc*100)+"%"
[cnn.txt](https://github.com/Hmy-bilibili/-cnn-/files/7109208/cnn.txt)
)
